src/cow-react/modules/trade/hooks/setupTradeState/useSetupTradeState.ts

```diff
@@ -1,195 +1,168 @@
-      if (chainIdFromProviderWasChanged) {
-        return providerChainId
-      } else {
-        // When wallet is not connected, then use chainId from URL by priority and fallback to provider's value
-        return validChainIdFromUrl
-      }
-    } else {
-      // When wallet is connected, then always use chainId from provider
-      return providerChainId
-    }
-  }, [account, chainIdFromProviderWasChanged, chainIdFromUrl, currentChainId])
+    const defaultState = getDefaultTradeState(currentChainId)
 
-  const updateStateAndNavigate = useCallback(() => {
-    if (!state) return
+    // SPECIAL CASE
+    if (isWalletConnected && providerAndUrlChainIdMismatch && prevTradeStateFromUrl) {
+      setRememberedUrlState(tradeStateFromUrl)
+      tradeNavigate(prevTradeStateFromUrl.chainId, prevTradeStateFromUrl)
+      console.debug(
+        '[TRADE STATE]',
+        'Remembering a new state from URL while changing chainId in provider',
+        tradeStateFromUrl
+      )
 
-    // Reset state to default when chainId was changed in the provider
-    const newState: TradeState = chainIdFromProviderWasChanged
-      ? getDefaultTradeState(newChainId)
-      : {
-          chainId: newChainId,
-          recipient: tradeStateFromUrl.recipient || state.recipient,
-          ...getUpdatedCurrenciesIds(tradeStateFromUrl, state),
-        }
+      return
+    }
 
-    console.debug('UPDATE TRADE STATE:', newState)
+    if (chainIdIsNotSupported || sameTokens || tokensAreEmpty || onlyChainIdIsChanged) {
+      tradeNavigate(currentChainId, defaultState)
+
+      if (chainIdIsNotSupported) {
+        console.debug('[TRADE STATE]', 'Url contains invalid chainId, resetting')
+      } else if (sameTokens) {
+        console.debug('[TRADE STATE]', 'Url contains invalid tokens, resetting')
+      } else if (tokensAreEmpty) {
+        console.debug('[TRADE STATE]', 'Url does not contain both tokens, resetting')
+      } else if (onlyChainIdIsChanged) {
+        console.debug('[TRADE STATE]', 'Only chainId was changed in URL, resetting')
+      }
 
-    updateState?.(newState)
+      return
+    }
 
-    tradeNavigate(newState.chainId, {
-      inputCurrencyId: newState.inputCurrencyId || null,
-      outputCurrencyId: newState.outputCurrencyId || null,
-    })
-  }, [tradeNavigate, newChainId, chainIdFromProviderWasChanged, state, updateState, tradeStateFromUrl])
+    updateState?.(tradeStateFromUrl)
+    console.debug('[TRADE STATE]', 'Applying a new state from URL', tradeStateFromUrl)
 
-  /**
-   * STEP 1
-   * Unlock network switching in the provider when chainId was changed in the URL
-   */
-  useLayoutEffect(() => {
-    if (chainIdFromUrlWasChanged) {
-      setIsChainIdSet(false)
-    }
-  }, [chainIdFromUrlWasChanged])
+    // Triggering only on changes from URL
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [tradeStateFromUrl])
```

Comment:
Does this work? (haven't tested it yet)
In the hook `currentChainId` and `rememberedUrlState` are relied upon but they are not part of the hook dependencies.

Reply:
Yes, it works